BGP EVPN vrf netns backend

pguibert6WIND opened the pull request:
BGP EVPN Route type 5 for vrf netns backend;

pguibert6WIND commented:
Hi @srimohans , @chiragshah6 ,

I also have a set of changes for EVPN RT2.
do you think this could be reviewed on the same PR ?

pguibert6WIND commented:
ci:rerun

pguibert6WIND commented:
Hi, any feedback please?

chiragshah6 commented on zebra/zebra_netns_notify.c:
```diff
@@ -72,7 +72,8 @@ static void zebra_ns_notify_create_context_from_entry_name(const char *name)
 	int ret;
-	ns_id_t ns_id, ns_id_external;
+	ns_id_t ns_id, ns_id_external, ns_id_relative = NS_UNKNOWN;
+	struct ns *def;
```
please rename def to more relevant

pguibert6WIND replied:
ok. will do it : default_ns

pguibert6WIND commented on zebra/zebra_l2.h:
```diff
@@ -37,6 +37,7 @@ extern "C" {
 	ifindex_t bridge_ifindex; /* Bridge Master */
 	struct interface *br_if;  /* Pointer to master */
+	ns_id_t ns_id; /* network namespace where bridge is */
 };
```
bridge must belong to the vrf, when used with vrf backend with namespace.
during my test, at one moment, you have the vxlan interface and you look for the bridge interface. having the index is not enough, I think. this is why I added the ns_id, so that the bridge can be found for both vrf backends

pguibert6WIND commented on zebra/zebra_ns.h:
```diff
@@ -72,6 +72,14 @@ int zebra_ns_final_shutdown(struct ns *ns);
+						 void *param_in,
+						 void **param_out),
+			void *param_in,
+			void **param_out);
```
wrong. indentation is correct, paarm in and param out are params number 2 and 3, and first param is a handled called exec_for_each_zns.

pguibert6WIND commented on zebra/zebra_vxlan.c:
```diff
@@ -1196,7 +1196,7 @@ static void zvni_print_mac(zebra_mac_t *mac, void *ctxt, json_object *json)
 			ifindex = mac->fwd_info.local.ifindex;
-			zns = zebra_ns_lookup(NS_DEFAULT);
+			zns = zebra_ns_lookup(mac->fwd_info.local.ns_id);
 			ifp = if_lookup_by_index_per_ns(zns, ifindex);
```
I am sorry I dont get what you mean by indentation here.
can you be more specific, pleas?

chiragshah6 replied:
I was referring to follow clang-format or to run tools/checkpatch.pl to align the indentation.

pguibert6WIND commented on zebra/zebra_ns.c:
```diff
@@ -217,3 +221,25 @@ int zebra_ns_config_write(struct vty *vty, struct ns *ns)
+						 void **param_out),
+			void *param_in,
+			void **param_out)
+{
```
wrong. indentation is correct, paarm in and param out are params number 2 and 3, and first param is a handled called exec_for_each_zns.

pguibert6WIND commented:
Hi @chiragshah6 , @srimohans , after a few polishing, you can have a look at the pull request.
thanks,

pguibert6WIND commented:
Hi @chiragshah6 , @srimohans , sorry to bother you.
I ping again for this pr.
thanks

pguibert6WIND commented:
Hi @chiragshah6 , @srimohans , sorry to bother you.
I ping again for this pr.
thanks

pguibert6WIND commented:
Hi @chiragshah6 , @srimohans , @donaldsharp , sorry to bother you.
I ping again for this pr.
thanks

pguibert6WIND commented:
Hi @chiragshah6 ,
thanks so much for reviewing it.
- documentation is added
- macvlan code change is supported.
I think it should be good to add a mac vlan config example too.

I let the others review it.
thanks

srimohans commented:
@pguibert6WIND I started reviewing this. Will finish in couple of days.

Could you please list the test cases that you had used to verify this functionality ? Also Please list test cases that are covered by topotest cases for this feature.

pguibert6WIND commented:
Hi Sri Mohan,
thanks for this.

This is not really a feature request, as one can see this as a bug fix when vrf backend is netns: EVPN was not working with that vrf backend.

> Could you please list the test cases that you had used to verify this functionality ? 

https://github.com/FRRouting/frr/pull/5077/commits/45542b35e57f8df02c28b6847b9019cdd3f784e1

indicates how you should configure your EVPN setup.
based on this, you can run BGP EVPN on the main vrf, enable advertise-all-vni. Automatically, the VNIs will be discovered across al the network namespaces.

With this change, you can bridge traffic or route traffic, if you have bridge vxlan interface in a network namespace different from the default one.

> Also Please list test cases that are covered by topotest cases for this feature.

There are no topotests. Actually, there are no test cases for EVPN in standard, but if there were some EVPN topotests based on standard vrf backend, I would be happy to handle the converstion to support vrf netns backend.

srimohans commented on zebra/zebra_vxlan.c:
```diff
@@ -5095,13 +5125,39 @@ struct interface *zl3vni_map_to_vxlan_if(zebra_l3vni_t *zl3vni)
+
+
+		zl3vni->local_vtep_ip = vxl->vtep_ip;
+		if (_pifp)
```
Do you mean _pifp or ifp here ?

srimohans commented on zebra/zebra_l2.c:
```diff
@@ -241,19 +243,22 @@ void zebra_l2if_update_bridge_slave(struct interface *ifp,
 	assert(zif);
 
+	zvrf = zebra_vrf_lookup_by_id(ifp->vrf_id);
+	assert(zvrf);
```
Should this be an assert or a simple return on zvrf being NULL ?

pguibert6WIND replied:
this call comes from netlink, so I don't think we can have cases where zvrf is not found.
however, to be safe, I will change the check.

pguibert6WIND commented on zebra/if_netlink.c:
```diff
@@ -495,6 +496,15 @@ static int netlink_extract_vxlan_info(struct rtattr *link_data,
 			*(struct in_addr *)RTA_DATA(attr[IFLA_VXLAN_GROUP]);
 	}
 
+	if (!attr[IFLA_VXLAN_LINK]) {
```
I don't think. If I remember well,vxlan_link appears only when interface is moved to an other network namespace. that means that this flag will not be present when you create a vxlan interface with the link on the same network namespace. we don't have to notify any error here.

pguibert6WIND commented on zebra/zebra_vxlan.c:
```diff
@@ -3620,51 +3620,35 @@ static zebra_vni_t *zvni_map_vlan(struct interface *ifp,
-	zif = br_if->info;
+	if (!in_param)
+		return ZNS_WALK_STOP;
+	br_if = in_param->br_if;
```
it is not possible as the calling function does not continue (line 3695) if br_if not null.
however, I add an assert, just to be sure.

pguibert6WIND commented on zebra/zebra_vxlan.c:
```diff
@@ -3839,16 +3840,51 @@ static struct interface *zvni_map_to_svi(vlanid_t vid, struct interface *br_if)
 {
-	struct zebra_ns *zns;
-	struct route_node *rn;
 	struct interface *tmp_if = NULL;
```
I think a pointer of pointer var2 must reference a pointer var1, otherwise var2 is not valid for being used.
in case you have doubt, we can ask to someone else.

srimohans replied:
I am good with your modifications.

pguibert6WIND commented on zebra/zebra_vxlan.c:
```diff
@@ -3839,16 +3840,51 @@ static struct interface *zvni_map_to_svi(vlanid_t vid, struct interface *br_if)
 	return tmp_if;
 }
 
+static int zvni_map_to_macvlan_ns(struct ns *ns,
```
I did not do it because natively, the code was already duplicated.
so here is the list:
(1)- zvni from svi => look for vxlan interface with same bridge that is not vlan awre
(2)- zvni map vlan =>  look for vxlan interface with same bridge and vlan id 
(3)- zvni map svi => look for vlan interface with same vlan id
(4)- zvni map mac vlan => look for macvlan interface with same link interface

it is possible to do a merge of some functions:
- (1) and (2)
- eventually apply to all ( include (3) and (4))

I tend to think it is better to not do this as it may bring complexity to maintain this single function.
however, thanks to confirm me you think it is better to do this, please.

pguibert6WIND replied:
Sri, any feedback ?

srimohans replied:
I am specifically referring to zvni_map_to_svi, zvni_map_to_svi_ns, zvni_map_to_macvlan and zvni_map_to_macvlan_ns functions. My earlier comment may not be very clear with respect to this.

pguibert6WIND commented on zebra/zebra_netns_notify.c:
```diff
@@ -218,9 +218,10 @@ static int zebra_ns_ready_read(struct thread *t)
 	}
-	if (err < 0)
+	if (err < 0) {
+		err = ns_switchback_to_initial();
```
sorry, this commit brings confusion.
initially, the problem was related to traces, and is not related to evpn.

I can see when I switch to/ from network namespaces to do some operations.
And I was seeing that I was entering twice on the same network namespace without leaving it.

So I proposed that patch. And when I read it again, I think this is not optimised. ( frr_with_privs is not present for instance).
so, if you don't mind, I will remove the PR.

srimohans replied:
Please go ahead.

pguibert6WIND commented:
Hi @srimohans , any feedback to the questions I raised, please?

pguibert6WIND commented:
I added a topotest with EVPN netns.
note that both ipv4 and ipv6 are tested, and ipv6 seems to fail in upstream for both kind of vrf, I would like to understand why.
here is the output of show ip route vrf <>
```
B>r 2001:102::21/128 [200/0] via ::ffff:c0a8:6415, bridge-101 onlink, 00:00:29
```
and here are the logs produced
```
2020/03/03 10:37:28 ZEBRA: netlink_route_multipath(): RTM_NEWROUTE 2001:101::41/128 vrf 1(254)
2020/03/03 10:37:28 ZEBRA: netlink_route_multipath() (single-path): nexthop via ::ffff:c0a8:6429  if 4(1)
2020/03/03 10:37:28 ZEBRA: netlink_talk: netlink-dp (NS 1) type RTM_NEWROUTE(24), len=84 seq=17 flags 0x401
2020/03/03 10:37:28 ZEBRA: [EC 4043309093] netlink-dp (NS 1) error: No route to host, type=RTM_NEWROUTE(24), seq=17, pid=3623726657
2020/03/03 10:37:28 ZEBRA: 1:192.168.101.41/32 update_from_ctx(): no fib nhg
2020/03/03 10:37:28 ZEBRA: 1:192.168.101.41/32 update_from_ctx(): rib nhg matched, changed 'true'
2020/03/03 10:37:28 ZEBRA: 1:192.168.101.41/32: Redist update re 0xdb5260 (bgp), old 0x0 (None)
2020/03/03 10:37:28 ZEBRA: 1:2001:101::41/128: Route install failed
```

anyone know why ipv4 mapped ipv6 nexthop does not work here?

pguibert6WIND commented:
@srimohans , @donaldsharp , topotest added for evpn vxlan.

pguibert6WIND commented:
@srimohans , @donaldsharp ,
thanks for finalising review, please.
this pr becomes very old, now :-)

pguibert6WIND commented:
Thanks Sri for approving.
Hi @donaldsharp , do you mind if you can give your go for that pr, please?

pguibert6WIND commented:
ci:rerun

pguibert6WIND commented:
ci:rerun

pguibert6WIND commented:
Ci:rerun

ton31337 commented on doc/user/bgp.rst:
```diff
@@ -2348,6 +2348,27 @@ the same behavior of using same next-hop and RMAC values.
+   ip netns exec vrf1 brctl addif bridge101 vxlan101
+
+This makes possible to separate not only layer 3 networks like VRF-lite networks.
+Also, VRF netns based make possible to separate layer 2 networks on separata VRF
```
typo here `separata`

pguibert6WIND replied:
ok

pjdruddy commented:
this is going to make rebasing evpn-mh hard - should we consider Double
committing there? @Anuradha Karuppiah <anuradhak@cumulusnetworks.com> @Quentin
Young <qlyoung@cumulusnetworks.com>

On Tue, May 19, 2020 at 7:13 AM Donatas Abraitis <notifications@github.com>
wrote:

> Merged #5077 <https://github.com/FRRouting/frr/pull/5077> into master.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FRRouting/frr/pull/5077#event-3350236834>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDGSX2XA7TCADL3YZ4SKNLRSIPOXANCNFSM4I3GYHYA>
> .
>
